calico_kubernetes/calico_kubernetes.py

```diff
@@ -147,6 +162,130 @@ def _configure_interface(self):
         print('Finished configuring network interface')
         return ep
 
+    def _container_add(self, container_name, workload_id, pid, ip, interface,
+                       hostname, orchestrator_id):
+        """
+        Add a container (on this host) to Calico networking with the given IP.
+
+        :param container_name: The name or ID of the container.
+        :param ip: An IPAddress object with the desired IP to assign.
+        :param interface: The name of the interface in the container.
+        """
+        # Check if the container already exists. If it does, exit.
+        try:
+            _ = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=orchestrator_id,
+                workload_id=workload_id
+            )
+        except KeyError:
+            # Calico doesn't know about this container.  Continue.
+            pass
+        else:
+            print("This container has already been configured with Calico Networking.")
+            sys.exit(1)
+
+        # Obtain information from Docker Client and validate container state
+        self._validate_container_state(container_name)
+
+        # Assign ip address through IPAM Client
+        try:
+            _ = self._datastore_client.assign_address(None, ip)
+        except PoolNotFound:
+            print "ERROR: IP address %s does not belong to any configured pools."\
+                  " Exiting." % ip
+            sys.exit(1)
+        else:
+            if _ is False:
+               print "ERROR: Failed to assign IP address %s. Exiting." % ip
+
+        # Create Endpoint object
+        try:
+            ep = self._datastore_client.create_endpoint(hostname, orchestrator_id,
+                                                        workload_id, [ip])
+        except AddrFormatError:
+            print "This node is not configured for IPv%d. Unassigning IP "\
+                  "address %s then exiting."  % ip.version, ip
+            self._datastore_client.unassign_address(None, ip)
+            sys.exit(1)
+
+        self._datastore_client.set_endpoint(ep)
+
+        # Create the veth, move into the container namespace, add the IP and
+        # set up the default routes.
+        # Update Endpoint in felix
+        ep.mac = ep.provision_veth(pid, interface)
```

Comment:
What happens if this fails?  I think we need to:

1) Not update the endpoint
2) unassign the address.